### ISSUE TITLE
Node connection lost after network change

### ISSUE THREAD

**Issue by geokyr** [fedbiomed/fedbiomed#1227]
I've been playing around with Fedbiomed and using it to try a few things about federated learning with health data. 

Currently I have a VM in the cloud that is running the vpnserver and the researcher containers and locally I'm running a node container and setting up fedbiomed to use the vpn deployment. Everything is going ok and the experiments are running and training on the node etc. But, when I switch for example from Wi-Fi to my 5G Hotspot, then the reseacher doesn't find my node.

I was previously having trouble when coming back to my laptop after it sleeping or shutting down, where the connection wouldn't work as well, but I found out on GitHub issues one thread where it basically said that the nohup start needs to run again. That fixed that problem. But for the one I described (where I change my network connection) I haven't found a solution.

I was thinking that the way Wireguard is working, from what I've searched, it should still be working (the keys and the persisent keep alive setting help). But it isn't working. Same thing happened when I switched off my auto dhcp IP, and then switched it back on (my IP most likely changed). The researcher can't find the node, even though nothing else has changed, except for me changing network connection. Do you have any idea what might be causing this? And any possible workarounds?

**Comment by sharkovsky** [comment]
Hi George, 

thank you for using Fed-BioMed! 

I will let @mvesin give you a more complete answer, since he is our Wireguard expert, but my feeling is that you're on the right track when you talk about IP. 
AFAIK, the Wireguard server (the `vpnserver` running in the cloud) saves the public IP of the clients after their first connection, and later on refuses the connection if it is coming from a different IP. 

If you are in a testing phase and can play around without fear that you might mess up your vpn (and therefore require a fresh deployment from scratch), you can test out the options below. If you cannot or do not want to perform a new deployment, then please be very cautious and wait for @mvesin's answer!

One option would be to manually update the connection by changing the wireguard config file. I don't have much time now but if you would potentially feel comfortable manually editing that file, I can give a closer look in the afternoon and provide more guidance. 

A second option could be to run the [`reresolve-dns.sh`](https://github.com/WireGuard/wireguard-tools/blob/master/contrib/reresolve-dns/reresolve-dns.sh ) script provided by the wireguard tools repository. I have no experience with their tools, so I can't really help you with running it (it looks like you only need to run the script on the server, without the need for arguments)

**Comment by geokyr** [comment]
Hello Francesco,

Thank you for your answer. I looked around in the configs etc and didn't find any indication that the vpnserver would save the client's IP somewhere, but all my experiments pointed to such a behavior.

I can play around in some development deployments we are testing now. 

> One option would be to manually update the connection by changing the wireguard config file.

Here I'm guessing you mean a wireguard config file that saves the IPs of the clients and not the wireguard config files that we sent from the server to the node, right? If so, tell me more when you have time. However, that still feels like its a bit too "complex" to be done manually each time, so I'll also look around the second option you gave, that seems promising.

Thank you and let me know when you have the time. I'll wait for @mvesin's answer as well.

**Comment by sharkovsky** [comment]
Yeah, do you have a file called `/config/wireguard/wg0.conf` in your `vpnserver` container? 
That file usually saves the local IP of each client (i.e. the IP inside the vpn network). Sometimes, it also saves the public IP of the client, in a field called `Endpoint` (if I remember correctly). 
Could you check if something similar is happening to you?

**Comment by geokyr** [comment]
I have this file. The content is:

```
[Interface]
ListenPort = 51820
PrivateKey = <private>

[Peer]
PublicKey = <public1>
PresharedKey = <preshared1>
AllowedIPs = 10.222.0.2/32
Endpoint = <ip>:36586

[Peer]
PublicKey = <public2>
PresharedKey = <preshared2>
AllowedIPs = 10.221.0.2/32
```

I also digged up the config that was sent to the node

```
# fedbiomed:
export FEDBIOMED_NODEID=NODETAG_67065d57c001987fcd1bb462
export FEDBIOMED_NET_IP=10.220.0.1

# client:
export VPN_IP=10.221.0.2
export VPN_SUBNET_PREFIX=14
export USE_WG_KERNEL_MOD=false

# server:
export VPN_SERVER_ENDPOINT=<ip>:<port2>
export VPN_SERVER_ALLOWED_IPS=10.220.0.0/14
export VPN_SERVER_PUBLIC_KEY=<public3>
export VPN_SERVER_PSK=<preshared2>
```

The field `public2` is equal to the key that the node sent back after configuration and the field `preshared2` is equal to the `VPN_SERVER_PSK` as seen above.

With this in mind, I guess that the node section is the second [Peer] one where I don't see an Endpoint field. The [Peer] section with the Endpoint field probably is refering to something else (maybe the researcher container?) because it's using the same IP as the machine that runs the vpnserver (researcher runs there as well), but on a different port that I don't see elsewhere.

**Comment by mvesin** [comment]
Hi George,

Thank you @sharkovsky for your answer. 
I would just add one point: if you plan to modify the `wg0.conf` files, please be sure to **stop the component** (node, researcher, vpnserver) whose `wg0.conf` you modify. `wg0.conf` is saved when the component stops, and is read when the component starts.

To be honest, we did not test your case recently, as our base test case uses hospitals as nodes, with "fixed" machines.
I would expect the following to work if the adress of your node changed:
- stop the node component on your node machine
- stop the vpnserver component on your vpnserver machine
- start the vpnserver component on your vpnserver machine
- start the node component on your node machine 
(the glitch may be that all components then need to be stopped/started)

To stop/start the components please follow https://fedbiomed.org/latest/user-guide/deployment/deployment-vpn/#misc-server-management-commands (only stop/start, don't clean configuration ...)

FYI stop/start components will not lose your component configurations. They are purposely saved in files outside of the docker image, to be able to have long lived Fed-BioMed components (they can be relaunched eg after a machine reboot).

Does it solve the issue regarding node PI change ?

If not, I may have the opportunity to test this case again next week on a testbed that I will (re)build.

**Comment by geokyr** [comment]
Hello Marc,

Thank you for your answer. I will try that as soon as possible and come back. I just want to ask, the solution you proposed with the stop node > stop vpnserver > start vpnserver > start node, is including the modification of the `wg0.conf` file? If yes, what is it exactly I need to modify, considering the sample file I posted earlier?

On another note, do you have any idea about the reresolved-dns script? I will test this in a bit as well, to see if it can automatically resolve the missing connection.

**Comment by sharkovsky** [comment]
> I just want to ask, the solution you proposed with the stop node > stop vpnserver > start vpnserver > start node, is including the modification of the wg0.conf file? If yes, what is it exactly I need to modify, considering the sample file I posted earlier?

No, Marc's proposed solution does not require any modification of the `wg0.conf` file. The idea (we hope!) is that the vpn server will automatically pick up the new IP, as long as you stop/restart them in the correct order. 

Please keep us updated!

**Comment by geokyr** [comment]
So I tried running the reresolve-dns script from wireguard-tools. It is running indefinitely. I changed the config path and tried running it from both outside the vpnserver container on the host machine and inside, with the proper paths each time.

About the other solution, I tried stop node (docker container) > stop vpnserver (script) > start vpnserver (script) > start node (docker container). The thing is that on our node side we are using docker and some uploaded images (fedbiomed-node) that we deploy on machines, where we might not have access to the host machine (we use another docker container to start the node container). So we can't run the stop and start scripts there since there is no fedbiomed installation on the host machine. However, the images and the commands that set up the creation/deletion are done in a way to replicate what is being done through the scripts. For example we run the nohup commands etc.

In that case, simply running that chain of operations didn't work, after changing from Wi-Fi to another Wi-Fi network. The researcher couldn't see the node.

I know that sounds complicated, everything is working out alright except for that part. We've also taken care of possible node container restarts and run the initialization (nohup etc) again. But when the connection (and possibly IP) changes, we can't find anything that recovers the connection.

I'm also tagging @NikosVlahakis, who is working with me at this problem.

**Comment by mvesin** [comment]
Hi again @geokyr 

I finally had the opportunity of testing on a setup consisting 
- a researcher on a server (Linux Fedora 38)
- 2 "fixed" nodes (Linux Fedora 38)
- a "mobile" node that roams from a mobile operator (connected using smartphone's Wi-Fi mobile hotspot) to an entreprise Wi-Fi (using difference Internet provider, completely different IP network prefixes and settings)

Using current branch `develop` thus "pre-v5.4.0" (but there should not be any difference with `v5.x` releases since they use gRPC transport also).

All components in this setup use Linux Fedora 38, but there should not be difference with most other Linux distributions.

Roamed several times back and forth mobile operator <=> enterprise network.

I did not try roaming *during* a training round operation though. This may introduce several seconds of network outage which may cause node <=> researcher messages to timeout and current request to fail. 
Please note this case is not really in our base case, as nodes in Fed-BioMed are primarily expected to be hospitals, thus with mostly fixed machines, where a roaming scenario would not occur (or not occur that often).

In all my tests **the Fed-BioMed node is up and running without any operation** as soon as it reconnects to another network.
No container relaunch, no Fed-BioMed node relaunch.
Just launch another operation from the researcher (eg launch a notebook), the node answers and participates in the training.

So I suspect the problem you observe comes from some limitation of your OS and/or docker support on your OS.

Is the "mobile" node using Windows with WSL ? MacOS ?

**Comment by geokyr** [comment]
Hello @mvesin,

Thank you for your time and testing.

It appears that once again the culprit is Windows+WSL, as this is what I had been using for my tests for the node side. Researcher and vpnserver were on Linux cloud machines.

It makes sense that the extra virtualization layer and the virtual networks and interfaces of WSL could cause an issue in the communication, and to be honest I expected it to work as it did in your tests, considering the way WireGuard works, with the keys, keepalive, etc.

If I get the chance to test this on a Linux machine on the node side, I'll come back and let you know.

**Comment by sharkovsky** [comment]
This issue seems resolved, closing it in the context of some cleanup. 
Feel free to reopen if you still have similar issues!
